spark-plugins/src/main/java/co/cask/hydrator/plugin/spark/FileStreamingSource.java

```diff
@@ -0,0 +1,225 @@
+                          TextInputFormat.class, filter, false)
+      .flatMap(new FormatFunction(conf.format, conf.schema));
+  }
+
+  /**
+   * Doesn't filter any files.
+   */
+  private static class NoFilter implements Function<Path, Boolean> {
+    @Override
+    public Boolean call(Path path) throws Exception {
+      return true;
+    }
+  }
+
+  /**
+   * Filters out files that don't have one of the supported extensions.
+   */
+  private static class ExtensionFilter implements Function<Path, Boolean> {
+    private final Set<String> extensions;
+
+    ExtensionFilter(Set<String> extensions) {
+      this.extensions = extensions;
+    }
+
+    @Override
+    public Boolean call(Path path) throws Exception {
+      String extension = Files.getFileExtension(path.getName());
+      return extensions.contains(extension);
+    }
+  }
+
+  /**
+   * Transforms kafka key and message into a structured record when message format and schema are given.
+   * Everything here should be serializable, as Spark Streaming will serialize all functions.
+   */
+  private static class FormatFunction implements FlatMapFunction<Tuple2<LongWritable, Text>, StructuredRecord> {
+    private final String format;
+    private final String schemaStr;
+    private transient Schema schema;
+    private transient RecordFormat<StreamEvent, StructuredRecord> recordFormat;
+
+    FormatFunction(String format, String schemaStr) {
+      this.format = format;
+      this.schemaStr = schemaStr;
+    }
+
+    @Override
+    public Iterable<StructuredRecord> call(Tuple2<LongWritable, Text> in) throws Exception {
+      // first time this was called, initialize schema and time, key, and message fields.
+      if (recordFormat == null) {
+        schema = Schema.parseJson(schemaStr);
+        FormatSpecification spec = new FormatSpecification(format, schema, new HashMap<String, String>());
+        recordFormat = RecordFormats.createInitializedFormat(spec);
+      }
+
+      StructuredRecord.Builder builder = StructuredRecord.builder(schema);
+      StructuredRecord messageRecord = recordFormat.read(new StreamEvent(ByteBuffer.wrap(in._2().getBytes())));
+      for (Schema.Field messageField : messageRecord.getSchema().getFields()) {
+        String fieldName = messageField.getName();
+        builder.set(fieldName, messageRecord.get(fieldName));
+      }
+      List<StructuredRecord> output = new ArrayList<>(1);
+      output.add(builder.build());
+      return output;
+    }
+  }
+
+  /**
+   * Configuration for the source.
+   */
+  public static class Conf extends ReferencePluginConfig {
+    private static final Set<String> FORMATS = ImmutableSet.of("text", "csv", "tsv", "clf", "grok", "syslog");
+
+    @Macro
+    @Description("The format of the source files. Must be text, csv, tsv, clf, grok, or syslog. Defaults to text.")
+    @Nullable
+    private String format;
+
+    @Description("The schema of the source files.")
+    @Nullable
```

Comment:
Why do we want schema to be nullable? If schema is not specified `conf.getSchema()` will throw NPE